src/vmspawn/vmspawn.c

```diff
@@ -760,6 +763,117 @@ static int start_virtiofsd(sd_bus *bus, const char *scope, const char *directory
+        if (!streq_ptr(pttype, "gpt")) {
+                log_debug("Didn't find a GPT partition table in '%s'.", arg_image);
+                return false;
+        }
+
+        errno = 0;
+        pl = blkid_probe_get_partitions(b);
+        if (!pl)
+                return log_debug_errno(errno_or_else(ENOENT), "Unable read partition table in '%s': %m", arg_image);
+
+        errno = 0;
+        nparts = blkid_partlist_numof_partitions(pl);
+        if (nparts == 0)
+                return log_debug_errno(errno_or_else(ENOENT), "Partition table in '%s' is empty: %m", arg_image);
+
+        sd_id128_t usr_partuuid = {};
+        for (int i = 0; i < nparts; i++) {
+                const char *p;
+                blkid_partition par;
+                sd_id128_t pt_parsed, uuid_parsed;
+
+                par = blkid_partlist_get_partition(pl, i);
+                if (!par) {
+                        log_debug("Failed to get partition %d in '%s', ignoring...", i, arg_image);
+                        continue;
+                }
+
+                r = blkid_partition_get_type_id128(par, &pt_parsed);
+                if (r == -ENXIO) {
+                        log_debug_errno(r, "Partition %d in '%s' has no type UUID, ignoring", i, arg_image);
+                        continue;
+                }
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to read type UUID of partition %d in '%s': %m", i, arg_image);
+
+                r = blkid_partition_get_uuid_id128(par, &uuid_parsed);
+                if (r == -ENXIO) {
+                        log_debug_errno(r, "Partition %d in '%s' has no UUID, ignoring", i, arg_image);
+                        continue;
+                }
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to read UUID of partition %d in '%s': %m", i, arg_image);
+
+                p = gpt_partition_type_uuid_to_string(pt_parsed);
+                if (!p) {
+                        log_debug("Couldn't determine type name for partition %d in '%s'.", i, arg_image);
+                        continue;
+                }
+
+                if (startswith(p, "root")) {
+                        root = strjoin("root=PARTUUID=", SD_ID128_TO_UUID_STRING(uuid_parsed));
+                        if (!root)
+                                return log_oom();
+
+                        *ret = TAKE_PTR(root);
+                        return 0;
+                }
+
+                if (startswith(p, "usr") && sd_id128_is_null(usr_partuuid)) {
+                        usr_partuuid = uuid_parsed;
+                }
+        }
+
+        if (!sd_id128_is_null(usr_partuuid)) {
+                root = strjoin("mount.usr=PARTUUID=", SD_ID128_TO_UUID_STRING(usr_partuuid));
+                if (!root)
+                        return log_oom();
+
+                *ret = TAKE_PTR(root);
+                return 0;
+        }
+
+        return -ENOENT;
+}
```

Comment:
please use dissect_image_file_and_warn() for this, and then use the data it returns. we want only a single dissector for this.

i.e. it will return a DissectedImage object to you, and you can get the part uuid from it via `.partitions[PARTITION_ROOT].uuid`.

Reply:
Fixed in eae47ee9d9.